vignettes/philosophy.Rmd

```diff
@@ -0,0 +1,163 @@
+It is also the default setting in the [`lintr`](https://github.com/jimhester/lintr/blob/master/README.md) package which we use to do the [static code analysis](https://en.wikipedia.org/wiki/Static_program_analysis).
+
+At first this seems a lot to memorise.
+RStudio make things easier when you activate all diagnostic options (_Tools_ > _Global options_ > _Code_ > _Diagnostics_).
+This highlights several problems by showing a squiggly line and/or a warning icon at the line number.
+Instead of learning all the rules by heart, run `check_lintr()` regularly and fix any issues that come up.
+Do this when working on every single project.
+Doing so enforces you to consistenly use the same coding style, making it easy to learn and use it.
+
+### Rules for coding style
+
+- `underscore_separated` names for functions, parameters and variables.
+- A line of code or comments must be no longer than 80 characters.
+  Pro tip: have RStudio display this margin in the editor.
+  _Tools_ > _Global options_ > _Code_ > _Display_ > _Show margin_
+- Object names must not be longer than 30 characters.
+- Start a new line
+    - after the pipe (`%>%`)
+    - never before but always after `{`
+    - before `}`
+- Use spaces instead of tabs.
+  Pro tip: make RStudio place 2 spaces when you hit the tab key.
+  _Tools_ > _Global options_ > _Code_ > _Editing_ > _Insert spaces for tabs_
+- Use spaces consistently
+    - Use exactly one space before and after
+        - assignments `<-`, `->`, `=`
+        - operators like `+`, `-`,`*`, `/`, ...
+    - No space before and one space after `,`
+    - One space before `(` except in `function(` (don't use `function (`)
+    - One space between `)` and `{`
+    - No space after or before `(` or `[`
+- Use double quotes (`"`) to define character strings.
+- No trailing whitespace
+    - spaces at the end of a line
+    - blank lines at the end of the script
+
+### Static code analysis checks
+
+- Is an object defined before you use it?
+- Do you use an object after you defined it?
+- Use `<-` or `->` to assign something.
+  Only use `=` to pass arguments to a function (e.g. `check_package(fail = TRUE)`).
+- Use `is.na(x)` instead of `x == NA`.
+- Use `seq_len()` or `seq_along()` instead of `1:length(x)`, `1:nrow(x)`, ...
+  When `length(x) == 0`, `1:length(x)` yields `c(1, 0)`, whereas `seq_along(x)` would yield an empty vector.
+- Don't store code in comments.
+  If you don't want to loose code, use version control systems like [`git`](https://git-scm.com/).
+- Avoid code with lots of nested loop or if statements.
+  If the code it too complex, you'll get a warning that the [cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity) is too high.
+  Tips for reducing the code complexity:
+    - Use `assertthat::assert_that()` to validate object or conditions instead of `if() stop()`.
+    - See if you can use `ifelse()` instead of `if()`.
+    - Split the main function of your code over a few sub functions.
```

Comment:
I'm not sure which is the best way to explain it.  Initially when using if, I always put the remaining part of the code in the else statement (resulting in a high complexity), but this is not always necessary, I learned later
```suggestion
    - Split the main function of your code over a few sub functions.
    - Don't use else if not strictly necessary.
```